### PR TITLE
Replace angebotschrifft layers by a filter

### PR DESCRIPTION
**Pull request by oterral**
# How to

Testable only inside geops VPN.

- Open [test app](https://deploy-preview-827--dev-trafimage-maps.netlify.app/?baselayers=ch.sbb.netzkarte,ch.sbb.netzkarte.luftbild.group&lang=de&layers=ch.sbb.construction.unterhalt.uebrige,ch.sbb.construction.unterhalt.bahnhof_strecke,ch.sbb.construction.unterhalt.strecke,ch.sbb.construction.unterhalt.bahnhof,ch.sbb.construction.ausbau.uebrige,ch.sbb.construction.ausbau.bahnhof_strecke,ch.sbb.construction.ausbau.strecke,ch.sbb.construction.ausbau.bahnhof&showFilter=true&x=928460.9&y=5908948.28&z=7)
- Open the menu 
- Select a diffrent filter for the layer Ausbau (select box "Filter by Angebotsschritt").
---> Clusters and points displayed should change when modifiying the filter.
 
# TODO
- [x] Remove dev links
- [x] Define default style of Ausbau layer
- [x] Test with trafimage on gitlab
- [x] Deploy a new react-spatial version with https://github.com/geops/react-spatial/pull/619

# Others

- [x] It's not a hack or at least an unauthorized hack :).
- [x] The images added are optimized.
- [x] Everything in ticket description has been fixed.
- [x] The author of the MR has made its own review before assigning the reviewer.
- [x] IE11 tested.
- [x] The title means something for a human being.
- [x] The title contains [WIP] if it's necessary.
- [x] Labels applied. if it's a release? a hotfix?
- [x] Tests added.

### CODE REVIEW

**Review by gevos — CHANGES_REQUESTED** [review]
I'm not yet finished with the review. But I wanted to give you my comments before the weekend.

- Where is the filtering for the visualization without clusters (single)
- I'm not sure if the filtering already works correctly. As a test, I have adapted the ConstructionPopup like in https://github.com/geops/trafimage-maps/pull/820/files. But, most popups do not contain the Angebotsschritt info although they are filtered.

File: src/components/TopicMenu/TopicMenu.js [hunk 1]
```diff
@@ -193,6 +194,27 @@ class TopicMenu extends PureComponent {
...
+                    <FiltersComponent layer={layer} />
+                  </div>
+                );
+              }
+              return null;
+              // {
+              //   /* layer.renderAfterChildren &&
+              //   layer.renderAfterChildren(layerTreeComp)}
+              //   <div>rencijmnrmn</div> */
+              // }
```
[the remark this answers is not in the document]
Comment (oterral) [reply src/components/TopicMenu/TopicMenu.js:216]: done

File: src/config/layers.js [hunk 2]
```diff
@@ -576,80 +577,25 @@ export const constructionDataLayer = new TrafimageMapboxLayer({
...
   visible: true,
+  mapboxLayer: constructionDataLayer,
   properties: {
     hasInfos: true,
     description: 'ch.sbb.construction.ausbau.group-desc',
+    filtersComponent: 'AusbauFilters',
   },
   children: [
-    angebotsSchritt2035,
+    // angebotsSchritt2035,
```
[the remark this answers is not in the document]
Comment (oterral) [reply src/config/layers.js]: done

File: src/layers/AusbauLayer/AusbauLayer.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+  applyNewFilter(filterToApply) {
+    const { mbMap } = this.mapboxLayer;
+    const style = mbMap.getStyle();
+    if (!mbMap || !style) {
+      return;
+    }
+
+    style.layers.forEach(({ id }) => {
+      // console.log(id, mbMap.getFilter(id));
```
[the remark this answers is not in the document]
Comment (oterral) [reply src/layers/AusbauLayer/AusbauLayer.js]: done

**Comment by oterral** [comment]
> 
> * Where is the filtering for the visualization without clusters (single)
> * I'm not sure if the filtering already works correctly. As a test, I have adapted the ConstructionPopup like in https://github.com/geops/trafimage-maps/pull/820/files. But, most popups do not contain the Angebotsschritt info although they are filtered.

The field is called 'angebotsschritt'

**Comment by gevos** [comment]
> * Where is the filtering for the visualization without clusters (single)
* This is still missing. When I zoom in to the single visualization, the red circle at visp is still visible for "Fertigstellung bis 2035" and "Projekte im Bau" although the data comes from "Fertigstellung 2030".
* Whereas in the the cluster visualization, both the "Bahnhof" checkbox and the "Fertigstellung 2030" can make the point at Visp invisible, this isn't true for the red circle in the single visualization.

> > * I'm not sure if the filtering already works correctly. As a test, I have adapted the ConstructionPopup like in https://github.com/geops/trafimage-maps/pull/820/files. But, most popups do not contain the Angebotsschritt info although they are filtered.
> 
> The field is called 'angebotsschritt'
* My fault. There is also the Unterhalt layer displayed.

Additionally, I was wondering why `showFilter=true` is not part in the permalink as default.

**Comment by oterral** [comment]
as discussed red point layers must be removed from the style.

@gevos can I merge ?

File: src/filters/AusbauFilters/AusbauFilters.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+import { useTranslation } from 'react-i18next';
+import {
+  Select,
+  MenuItem,
+  FormControl,
+  InputLabel,
+  makeStyles,
+} from '@material-ui/core';
+import AusbauLayer from '../../layers/AusbauLayer';
+import { updateConstructions } from '../../config/layers';
```
Comment (gevos) [inline src/filters/AusbauFilters/AusbauFilters.js:12]: Is it really good style to import a function from a config file?

File: src/layers/AusbauLayer/AusbauLayer.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+  applyNewFilter(filterToApply) {
+    const { mbMap } = this.mapboxLayer;
+    const style = mbMap.getStyle();
+    if (!mbMap || !style) {
+      return;
+    }
+
+    style.layers.forEach(({ id }) => {
+      // This filter must affect nur Ausbau layers.
```
Comment (gevos) [inline src/layers/AusbauLayer/AusbauLayer.js:38]:
```suggestion
      // This filter must affect only Ausbau layers.
```

File: src/layers/AusbauLayer/AusbauLayer.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
+import { MapboxStyleLayer } from 'mobility-toolbox-js/ol';
+import qs from 'query-string';
+
+/**
+ * Layer for ausbau
+ * Extends {@link https://mobility-toolbox-js.netlify.app/api/class/src/ol/layers/VectorLayer%20js~VectorLayer%20html}
```
Comment (gevos) [inline src/layers/AusbauLayer/AusbauLayer.js:6]: MapboxStyleLayer instead of VectorLayer

**Review by gevos — CHANGES_REQUESTED** [review]
Nearly approved. Just some small issues.

Additionally, I think, it would be better if the filter would end before the info buttons on the right side.

**Comment by gevos** [comment]
> > * Where is the filtering for the visualization without clusters (single)
> 
>     * This is still missing. When I zoom in to the single visualization, the red circle at visp is still visible for "Fertigstellung bis 2035" and "Projekte im Bau" although the data comes from "Fertigstellung 2030".
> 
>     * Whereas in the the cluster visualization, both the "Bahnhof" checkbox and the "Fertigstellung 2030" can make the point at Visp invisible, this isn't true for the red circle in the single visualization.
 
The red circle layer will be removed.
 
> Additionally, I was wondering why `showFilter=true` is not part in the permalink as default.

This is done by intention.

File: src/filters/AusbauFilters/AusbauFilters.js (hunk 4 above)
Comment (oterral) [reply src/filters/AusbauFilters/AusbauFilters.js:12]: Not really but I didn't want to break the cluster and haven't a better idea  right now.

File: src/layers/AusbauLayer/AusbauLayer.js (hunk 6 above)
Comment (oterral) [reply src/layers/AusbauLayer/AusbauLayer.js]: fixed

File: src/filters/AusbauFilters/AusbauFilters.js (hunk 4 above)
Comment (oterral) [reply src/filters/AusbauFilters/AusbauFilters.js:12]: maybe the cluster should have it's own class layer and then we just import the layer here and do {code}clusterLayer.refresh(){code}
but it needs more thinking. We will see in the future if we have more use case like this.

**Pull request merged**